Introduce ObjectCleaner and use it in FastThreadLocal to ensure FastThreadLocal.onRemoval(...) is called

normanmaurer opened the pull request:
Motivation:

    There is no guarantee that FastThreadLocal.onRemoval(...) is called if the FastThreadLocal is used by "non" FastThreacLocalThreads. This can lead to all sort of problems, like for example memory leaks as direct memory is not correctly cleaned up etc.

    Beside this we use ThreadDeathWatcher to check if we need to release buffers back to the pool when thread local caches are collected. In the past ThreadDeathWatcher was used which will need to "wakeup" every second to check if the registered Threads are still alive. If we can ensure FastThreadLocal.onRemoval(...) is called we do not need this anymore.

    Modifications:

    - Introduce ObjectCleaner and use it to ensure FastThreadLocal.onRemoval(...) is always called when a Thread is collected.
    - Deprecate ThreadDeathWatcher
    - Add unit tests.

    Result:

    Consistent way of cleanup FastThreadLocals when a Thread is collected.

normanmaurer commented:
There is still some TODO to fix ... Just wanted to show the idea

johnou commented:
@normanmaurer could the same approach be used with Recycler instead of finalize?

normanmaurer commented:
@johnou not sure what you refer to with finalize.... any pointers?

johnou commented:
@normanmaurer 

io.netty.util.Recycler.WeakOrderQueue#finalize
```java
        @Override
        protected void finalize() throws Throwable {
            try {
                super.finalize();
            } finally {
                // We need to reclaim all space that was reserved by this WeakOrderQueue so we not run out of space in
                // the stack. This is needed as we not have a good life-time control over the queue as it is used in a
                // WeakHashMap which will drop it at any time.
                Link link = head;
                while (link != null) {
                    reclaimSpace(LINK_CAPACITY);
                    link = link.next;
                }
            }
        }
```

normanmaurer commented:
Yes most likely

> Am 01.12.2017 um 17:20 schrieb Johno Crawford <notifications@github.com>:
> 
> @normanmaurer
> 
> io.netty.util.Recycler.WeakOrderQueue#finalize
> 
>         @Override
>         protected void finalize() throws Throwable {
>             try {
>                 super.finalize();
>             } finally {
>                 // We need to reclaim all space that was reserved by this WeakOrderQueue so we not run out of space in
>                 // the stack. This is needed as we not have a good life-time control over the queue as it is used in a
>                 // WeakHashMap which will drop it at any time.
>                 Link link = head;
>                 while (link != null) {
>                     reclaimSpace(LINK_CAPACITY);
>                     link = link.next;
>                 }
>             }
>         }
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ejona86 commented on common/src/main/java/io/netty/util/internal/AutomaticCleanupReference.java:
```diff
@@ -0,0 +1,94 @@
+                Thread cleanupThread = new Thread(new Runnable() {
+                    @Override
+                    public void run() {
+                        for (;;) {
```
So this runs forever? Isn't that _guaranteed_ to be a classloader leak via this Runnable?

ejona86 commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -30,38 +29,45 @@
+                CLEANER_RUNNING.set(false);
+
+                // Its important to first access the LIVE_SET and then CLEANER_RUNNING to ensure correct
+                // behavior in multi-htraded enviroments.
```
s/htraded/threaded/

ejona86 commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+     * to the object anymore, which typically happens once the {@link Thread} dies.
+     *
+     * This should only be used if there are no other ways to execute some cleanup once the {@link Thread} dies as
+     * its not a cheap way to handle the cleanup.
```
nit: s/its/it's/

ejona86 commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private ThreadCleaner() {
+        // Only contains a  static method.
```
nit: there's a double space

ejona86 commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+        // behavior in multi-threaded enviroments.
+        LIVE_SET.add(reference);
+
+        // Check if there is arleady
```
s/arleady/already/

ejona86 commented:
> I guess the failure case is the ThreadLocal instance could be thrown away, and let the storage evaporate (since Thread uses WeakReference for the mapping), but the Thread itself remain running.

So, this turned out to be semi-wrong. Or rather, if you leave a Thread running, "you're gonna have a bad time." I realized that even using `finalize()` you can't necessarily clean up resources if the Thread lives because ThreadLocal is only lazily cleaned up in the ThreadLocalMap. Any values held are strong references, and so they might be retained for the rest of the Thread's lifetime.

So I'm not trying to solve that any more :-). And that makes multiple solutions we've considered solvable. The only thing we need to be wary of is clearing the Thread.contextClassLoader.

johnou commented on buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:
```diff
@@ -440,14 +441,26 @@ protected synchronized PoolThreadCache initialValue() {
             if (useCacheForAllThreads || current instanceof FastThreadLocalThread) {
                 // If our FastThreadLocalThread will call FastThreadLocal.removeAll() we not need to use
                 // the ThreadDeathWatcher to release memory from the PoolThreadCache once the Thread dies.
```
update javadoc?

normanmaurer replied:
+1 ... Let me do this

johnou commented on common/src/main/java/io/netty/util/Recycler.java:
```diff
@@ -416,7 +416,10 @@ protected void finalize() throws Throwable {
+
+        // We store the Thread in a WeakReference as otherwise we may be the only ones that still hold a strong
+        // Reference to the Thread itself after it died.
+        final WeakReference<Thread> threadRef;
```
should this be in the same pull request?

normanmaurer replied:
I did it in the same as I noticed it here because the Thread was never GCed. Maybe I should do it in a separate PR tho..

normanmaurer replied:
@johnou https://github.com/netty/netty/pull/7499 ... Even added a test-case :)

carl-mastrangelo commented:
Out of curiosity: why not spawn the cleaner thread when the original thread goes away?    That would keep the number of active threads down.   

Also, why not do the cleanup in a `finalize()` method?   This is actually garbage collection, so it seems reasonable to do it there and not spawn another thread.

normanmaurer commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+ * Allows a way to register some {@link Runnable} that will executed once there are no references to the {@link Thread}
+ * anymore. This typically happens once the {@link Thread} dies / completes.
+ */
+public final class ThreadCleaner {
```
nope... `WeakReference` was good enough.

normanmaurer commented:
@carl-mastrangelo answers / comments below....


> Out of curiosity: why not spawn the cleaner thread when the original thread goes away, rather than when it is first attached? That would keep the number of active threads down.

How would you detect that the original Thread goes away ? Also I think its not so bad to have one extra thread that is doing the cleanup for everything. I will most likely use it in other places as well at some point.

> Also, why not do the cleanup in a finalize() method? This is actually garbage collection, so it seems reasonable to do it there and not spawn another thread.

I actually had a PR (#7446) with a `finalize()` at some point but @ejona86 asked me to not do this as `finalize()` is marked as `@deprecated` in Java9 and so should not be used anymore.

johnou commented:
@carl-mastrangelo Object.finalize is deprecated, the recommended replacements are PhantomReference and the new Cleaner class.

normanmaurer commented:
@johnou @ejona86 @carl-mastrangelo PTAL again as I changed the approach a bit in a followup commit that I pushed. So the "real problem" why we needed all of this is because of the user uses a non `FastThreadLocalThread` there is no guarantee that we will call `FastThreadLocal.removeAll()` at all once the Thread goes away. The cleanup of the `PoolThreadCache` is done in `FastThreadLocal.onRemove(...)` that will be triggered by the `removeAll()`.

Because of this I now moved the usage of `ThreadCleaner.register(...)` into the `FastThreadLocal` itself which has the affect that we are guaranteed that `onRemoval(...)` is always called (even if we use a non `FastThreadLocal`. I think this is a more general fix to the problem and also fixes a possible memory leak that could happen `NativeDatagramPacketArray` if non `FastThreadLocal` threads are used to access it as `onRemoval` would never been called before for these:

https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/java/io/netty/channel/epoll/NativeDatagramPacketArray.java#L43

WDYT ?

ejona86 commented on common/src/main/java/io/netty/util/concurrent/FastThreadLocal.java:
```diff
@@ -57,21 +58,25 @@ public static void removeAll() {
         }
     }
 
+    private static void removeAll(InternalThreadLocalMap threadLocalMap) {
```
Is there a reason this was split out into its own function?

I had expected the `Runnable` passed to `ThreadCleaner` would use it, but it uses `remove(...)` instead. It's a bit hard to determine what's the difference between the two methods.

normanmaurer replied:
Actually sorry it should use `removeAll` and I messed up :(

normanmaurer replied:
Sorry what I wanted to say was that we not need to factor out removeAll and can just call remove... will fix it.

ejona86 commented:
> Because of this I now moved the usage of ThreadCleaner.register(...) into the FastThreadLocal itself which has the affect that we are guaranteed that onRemoval(...) is always called (even if we use a non FastThreadLocal

It also means that you'll use the ThreadCleaner even if nothing stored in the FastThreadLocal needs cleaning. Normally that'd be rare, but if caching is disabled for non-FastThreadLocalThreads, then it'd be more frequent. Not sure it matters at all.

normanmaurer commented:
@ejona86 yeah I agree... That said I think it may be still the preferred way to fix it and ensure `onRemoval` is always called and so can depend on it... WDYT ?

carl-mastrangelo commented on common/src/main/java/io/netty/util/concurrent/FastThreadLocalThread.java:
```diff
@@ -89,4 +89,13 @@ public final void setThreadLocalMap(InternalThreadLocalMap threadLocalMap) {
+     */
+    @UnstableApi
+    public static boolean willCleanupFastThreadLocals(Thread thread) {
+        return thread instanceof FastThreadLocalThread && ((FastThreadLocalThread) thread)
```
might be clearer to break on `&&`

carl-mastrangelo commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+    // This will hold a reference to the ThreadCleanerReference which will be removed once we called cleanup()
+    private static final Set<ThreadCleanerReference> LIVE_SET = new ConcurrentSet<ThreadCleanerReference>();
+    private static final ReferenceQueue<Object> REFERENCE_QUEUE = new ReferenceQueue<Object>();
+    private static final AtomicBoolean CLEANER_RUNNING = new AtomicBoolean(false);
```
nit: caps are usually for constants.

normanmaurer replied:
I think CAPS are for things that are static final so I think this is fine. Sure the content can be changed but this is ok IMHO.

carl-mastrangelo commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,131 @@
+            cleanupThread.setName("ThreadCleanerReaper");
+
+            // This Thread is not a daemon as it will die once all references to the registered Threads will go away.
+            cleanupThread.setDaemon(false);
```
Daemon is for not blocking shutdown.  In a shutdown scenario, how important is it to clean up refs?

normanmaurer replied:
actually I think we could also just use a non-daemon thread as the thread will go away once there is no references anymore.

ejona86 commented:
> That said I think it may be still the preferred way to fix it and ensure `onRemoval` is always called and so can depend on it... WDYT ?

I'm fine with it. It might be unnecessary in some cases, but doesn't seem like it hurts much and does remove a type of bug-prone behavior.

normanmaurer commented:
@carl-mastrangelo PTAL again

Scottmitch commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,132 @@
+     * This should only be used if there are no other ways to execute some cleanup once the {@link Thread} dies as
+     * it is not a cheap way to handle the cleanup.
+     */
+    public static void register(Thread thread, Runnable cleanupTask) {
```
this class is `internal` ... consider:
1. exposing a method where the user can directly create a `ThreadCleanerReference` to avoid allocating a `Runnable` and `ThreadCleanerReference`
2. Make the class package private and you can use the same technique as above to reduce object allocation.

normanmaurer replied:
I am not concerned about this as this should be something that is done very frequently and I think this design is cleaner.

Scottmitch replied:
Agreed the current APIs are nicer. However it is internal and doesn't need to be generic right? We could potentially save 1 object allocation per thread local. Since it is internal we can always change it later but maybe nice to keep thread local overhead down if we don't have to be generic.

normanmaurer replied:
@Scottmitch as discussed offline I will merge it as it is for now and we can address this if we find out its an issue later on. The current design is nicer :)

Scottmitch replied:
sgtm ... internal == change later if necessary

normanmaurer replied:
exactly 👍

Scottmitch commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,132 @@
+                }
+            }
+            if (interrupted) {
+                // As we catched the InterruptedException above we should mark the Thread as interrupted.
```
`catched` -> `caught`

normanmaurer commented on common/src/main/java/io/netty/util/internal/ThreadCleaner.java:
```diff
@@ -0,0 +1,132 @@
+        LIVE_SET.add(reference);
+
+        // Check if there is already
+        if (CLEANER_RUNNING.compareAndSet(false, true)) {
```
again I think we should not care as this should not happen very frequently and the CAS is the thing that I would be concerned last about (as the other stuff will be the "expensive")

Scottmitch replied:
fair enough ... I guess even if folks are doing get/remove this may not buy much.

normanmaurer commented:
@Scottmitch yeah I think how we test here is the best we can do for now...

carl-mastrangelo commented:
cc: @zpencer

normanmaurer commented:
@mosesn you may be interested as well.

normanmaurer commented:
Cherry-picked into 4.1 (e329ca1cf3cecd41c9cf1a5a7be369a65f002ab5) and 4.0 (00ae6274c015804546f5f12d2b795ec4d105ee71)

zpencer commented on common/src/test/java/io/netty/util/concurrent/FastThreadLocalTest.java:
```diff
@@ -75,4 +75,61 @@ public void run() {
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                assertEquals(Thread.currentThread().getName(), threadLocal.get());
```
A bit late to this PR, but what happens if code sets a FastThreadLocal and the thread exits before it's read? It looks like the cleaner code is only registered on the `get` path.

normanmaurer replied:
@zpencer good point. Let me fix this.

normanmaurer replied:
@zpencer https://github.com/netty/netty/pull/7534 PTAL
